.github/workflows/scripts/lint-includes.py

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python3
+# Copyright 2020 Redpanda Data, Inc.
+#
+# Use of this software is governed by the Business Source License
+# included in the file licenses/BSL.md
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0
+
+import sys
+import pathlib
+import subprocess
+import re
+
+
+def get_modules(root):
+    """
+    Get new-style modules (ie those that use the new layout where public headers
+    are separated from private headers).
+    """
+    modules = (x for x in root.iterdir() if x.is_dir())
+    for module in modules:
+        if (module / "include").is_dir():
+            yield module
+
+
+def validate_module(module):
+    """
+    Validate new-style module.
+
+    1. Should contains only the `include/<module>` directory
+    2. Should not contain ambiguous public/private file paths
+
+    Returns public files
+    """
+    include_root = module / "include"
+    include_path = include_root / module.name
+    assert include_path.is_dir(), f"Missing module directory {include_path}"
+    assert len([x for x in include_root.iterdir()
+                ]) == 1, f"{include_root} should only contain module directory"
+
+    public_files = set()
+    private_files = set()
+    for path in module.rglob("*"):
+        if not path.is_file():
+            continue
+        if include_path in path.parents:
+            public_files.add(path.relative_to(include_path))
+        else:
+            private_files.add(path.relative_to(module))
+
+    assert public_files.isdisjoint(
+        private_files), f"{public_files.intersection(private_files)}"
+
+    # add back in the module name to the path like "<module>/header.h"
+    return set(module.name / p for p in public_files)
+
+
+def validate_modules(root):
+    """
+    Validate all new-style modules and return a collection of all files that are
+    in public include paths and all module names.
+    """
+    public_files = set()
+    module_names = set()
+    for module in get_modules(root):
+        module_names.add(pathlib.Path(module.name))
+        for f in validate_module(module):
+            assert f not in public_files
+            public_files.add(f)
+    return public_files, module_names
+
+
+def get_include_path(line):
+    m = re.match(r".*#include.*\<(.+)\>.*", line)
+    if m:
+        return pathlib.Path(m.group(1))
+    m = re.match(r".*#include.*\"(.+)\".*", line)
+    if m:
+        return pathlib.Path(m.group(1))
+
+
+def get_module_includes(root, module_names):
+    includes = set()
+    output = subprocess.check_output(f"git grep #include {root}".split())
+    for line in output.decode().split("\n"):
+        include_path = get_include_path(line)
+        if include_path is None:
+            continue
+        if len(include_path.parents) <= 1:
+            continue
+
+        include_module = include_path.parents[-2]
```

Comment:
This doesn't account for nested modules

Reply:
yeh, i didn't realize we had nested modules.

Reply:
handling this case now

```
Found module wasm/parser: src/v/wasm/parser
Validating module wasm/parser
Module root src/v/wasm/parser
Include root src/v/wasm/parser/include
Include path src/v/wasm/parser/include/wasm/parser
Found 1 exports: sample ['src/v/wasm/parser/include/wasm/parser/parser.h', '...']
Found 7 internal sample ['src/v/wasm/parser/CMakeLists.txt', 'src/v/wasm/parser/leb128.h', 'src/v/wasm/parser/tests/CMakeLists.txt', '...']

Found module random: src/v/random
```